framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,601 @@
+        private final PrintWriter printWriter;
+
+        /**
+         * Constructs a new ImportDefCollector, which will run on the given AScene when its {@code
+         * visit} method is called.
+         *
+         * @param scene the scene whose imported annotations should be printed
+         * @param printWriter the writer onto which to write the import statements
+         * @throws DefException if the DefCollector does not succeed
+         */
+        ImportDefCollector(AScene scene, PrintWriter printWriter) throws DefException {
+            super(scene);
+            this.printWriter = printWriter;
+        }
+
+        /**
+         * Write an import statement for a given AnnotationDef. This is only called once per
+         * annotation used in the scene.
+         *
+         * @param d the annotation definition to print an import for
+         */
+        @Override
+        protected void visitAnnotationDef(AnnotationDef d) {
+            if (!d.name.contains("+")) {
+                printWriter.println("import " + d.name + ";");
+            }
+        }
+    }
+
+    /**
+     * For a given class, this prints the hierarchy of outer classes, and returns the number of
+     * curly braces to close with. The classes are printed with appropriate opening curly braces, in
+     * standard Java style. This routine does not attempt to indent them correctly.
+     *
+     * <p>When an inner class is present in an AScene, its name is something like "Outer.Inner".
+     * Writing a stub file with that name would be useless to the stub parser, which expects inner
+     * classes to be properly nested.
+     *
+     * @param basename the simple name of the class (without the package), in fully-qualified form
+     *     (that is, without any {@code $}s)
+     * @param classname the fully-qualified name of the class in question, so that it can be printed
+     *     as an enum if it is one
+     * @param aClass the AClass for {@code classname}
+     * @param printWriter the writer where the class definition should be printed
+     * @return the number of outer classes within which this class is nested
+     */
+    private int printClassDefinitions(
+            String basename,
+            @FullyQualifiedName String classname,
+            AClass aClass,
+            PrintWriter printWriter) {
+
+        String nameToPrint = basename;
+        String remainingInnerClassNames = "";
+        if (basename.contains(".")) {
+            nameToPrint = basename.substring(0, basename.indexOf('.'));
+            remainingInnerClassNames = basename.substring(basename.indexOf('.') + 1);
+        }
+
+        // For any outer class, print "class".  For a leaf class, print "enum" or "class".
+        if ("".equals(remainingInnerClassNames) && enumConstants.containsKey(classname)) {
+            printWriter.print("enum ");
+        } else {
+            printWriter.print("class ");
+        }
+        formatAnnotations(aClass.tlAnnotationsHere);
```

Comment:
The return value of this method is ignored.  Can you add a test case?

Reply:
I've fixed this, but there's no way to write a test for it (and it won't ever be exercised), because WPI has no way to infer types for class declarations.